packages/shared/src/hooks/useUserSettingsLocalStorage.ts

```diff
@@ -0,0 +1,188 @@
+import * as React from 'react';
+
+// Default ODF user settings storage path
+const DEFAULT_ODF_CONSOLE_USER_SETTINGS_KEY = 'odf-console-user-settings';
+const STORAGE_EVENT = 'storage';
+
+export const deserializeData = (data: string | null) => {
+  if (typeof data !== 'string') {
+    return data;
+  }
+  try {
+    return JSON.parse(data);
+  } catch {
+    return data;
+  }
+};
+
+export const serializeData = <T>(data: T) => {
+  if (typeof data === 'string') {
+    return data;
+  }
+  try {
+    return JSON.stringify(data);
+  } catch {
+    return data;
+  }
+};
+
+// Get user settings from local storage
+const getDataFromLocalStorage = (storageKey: string) =>
+  deserializeData(localStorage.getItem(storageKey)) ?? {};
+
+// Initialize state
+const initializeDataFromLocalStorage = <T>(
+  storageKey: string,
+  keyRef: string,
+  defaultValue: T
+) => {
+  const valueInStorage = getDataFromLocalStorage(storageKey);
+  return valueInStorage?.hasOwnProperty(keyRef) &&
+    valueInStorage[keyRef] !== undefined
+    ? valueInStorage[keyRef]
+    : defaultValue;
+};
+
+// Update local state with user settings by event
+const syncLatestDataFromLocalStorage = <T>(
+  isMounted: boolean,
+  storageKey: string,
+  storage: Storage,
+  key: string,
+  newValue: string,
+  data: T,
+  userSettingsKey: string,
+  setData: React.Dispatch<any>
+) => {
+  if (isMounted && storage === localStorage && key === storageKey) {
+    const newData = deserializeData(newValue)?.[userSettingsKey];
+    if (
+      newData !== undefined &&
+      serializeData(newData) !== serializeData(data)
+    ) {
+      setData(newData);
+    }
+  }
+};
+
+const updateLocalStorage = <T>(
+  storageKey: string,
+  userSettingsKey: string,
+  localData: any,
+  newState: T
+) => {
+  // Trigger update also when unmounted
+  const dataToUpdate = {
+    ...localData,
+    ...{
+      [userSettingsKey]: newState,
+    },
+  };
+
+  const newValue = serializeData(dataToUpdate);
+
+  try {
+    // Update the local storage
+    localStorage.setItem(storageKey, newValue);
+
+    // Dispatch storage event to sync the settings changes for other tabs
+    generateStorageEventToUpdate(storageKey, newValue);
+  } catch (err) {
+    // eslint-disable-next-line no-console
+    console.error(
+      `Error while updating local storage for key ${storageKey}`,
+      err
+    );
+  }
+};
+
+// Generate storage event to sync latest settings on other tabs
+const generateStorageEventToUpdate = (storageKey, newValue: string) => {
```

Comment:
IIRC this is not for other tabs, but for current itself, right ??
```suggestion
const generateStorageEventToUpdate = (storageKey, newValue: string) => {
```

Reply:
in which case this comment is wrong...

Reply:
no need to block the PR on this, but we can do it in next follow-up PR...

Reply:
check note: https://developer.mozilla.org/en-US/docs/Web/API/Window/storage_event

Reply:
got it !

Reply:
Just FYI: https://github.com/red-hat-storage/odf-console/issues/1629